Add undocumented properties

bancer opened the pull request:
See https://book.cakephp.org/2.0/en/controllers/request-response.html#accessing-request-parameters
and https://api.cakephp.org/2.10/source-class-UpgradeShell.html#345-403

bancer committed:
Add undocumented properties

See https://book.cakephp.org/2.0/en/controllers/request-response.html#accessing-request-parameters
and https://api.cakephp.org/2.10/source-class-UpgradeShell.html#345-403

garas commented:
These properties are handled by magic `__get()`, `__set()`, `__isset()` methods. After you have added these properties, magic methods are no longer called, breaks behavior and fails tests.

Adding `@property` tags to class PhpDoc would help IDE to see these properties, but they don't allow mark properties as `@deprecated`.

bancer committed:
Initialize $params to null

bancer committed:
Document magic properties

bancer committed:
Document more magic properties, trigger deprecated notices

ADmad commented on lib/Cake/Controller/Controller.php:
```diff
@@ -389,15 +396,19 @@ public function __isset($name) {
 			case 'here':
 			case 'webroot':
 			case 'data':
+				trigger_error($message, E_USER_DEPRECATED);
```
New deprecation error messages shouldn't be added in bugfix releases.

bancer replied:
Would it be okay to write to a log file instead?

bancer committed:
Use $controller->request->params instead $controller->params

saeideng commented:
I'm :-1:  with adding any new deprecation/error messages into cake 2.x series 
CakePHP 2.x is in maintenance mode

bancer committed:
Write to a log file instead of trigger errors

chinpei215 commented:
Please do not write logs either. If we had a chance to release release 2.11, I would change `__get()` to `&__get()` though.

ADmad commented on lib/Cake/Controller/Controller.php:
```diff
@@ -389,15 +396,19 @@ public function __isset($name) {
 			case 'here':
 			case 'webroot':
 			case 'data':
+				CakeLog::write(E_USER_DEPRECATED, $message);
```
I don't think filling up the logs is going to do any good either. Lot of 2.x plugins / apps might never get updated to remove use of deprecated code.

bancer committed:
Remove logs

bancer committed:
Imrove documentation

bancer committed:
Remove unused variable

bancer committed:
Improve documentation, add type check.

It seems like $this->uses can also be of type false; however, parameter
$array of array_unshift() does only seem to accept array.

bancer committed:
Improve documentation

bancer committed:
Improve documentation

bancer committed:
Declare undeclared property

bancer committed:
Add extra type checks

bancer committed:
Adjust type check

bancer committed:
Improve documentation, initiate uninitialized variables.

dereuromark commented on lib/Cake/Controller/Component/RequestHandlerComponent.php:
```diff
@@ -72,6 +72,13 @@ class RequestHandlerComponent extends Component {
+/**
+ * Array of parameters parsed from the URL.
+ *
+ * @var array
```
|null missing

bancer committed:
Improve documentation, reset variable

saeideng commented on lib/Cake/basics.php:
```diff
@@ -227,6 +226,7 @@ function h($text, $double = true, $charset = null) {
 		}
 		if (is_string($double)) {
 			$charset = $double;
+			$double = true;
```
this is not similar to `master` version 
https://github.com/cakephp/cakephp/blob/21c8fce5ca8b68292b948a28606006faa935b579/src/Core/functions.php#L66-L71

bancer replied:
It is not clear what you mean by that. The fourth parameter to `htmlspecialchars` below must be a boolean. So, `$double` should be converted to a boolean as it is intended according to the method signature `function h($text, $double = true, $charset = null)`.

saeideng replied:
I meant you should fix it first in `master`

bancer replied:
https://github.com/cakephp/cakephp/pull/12719

saeideng commented on lib/Cake/View/View.php:
```diff
@@ -455,7 +455,7 @@ public function elementExists($name) {
  * the view will be located along the regular view path cascade.
  *
- * @param string $view Name of view file to use
+ * @param bool|string $view Name of view file to use.
```
use of `bool` for first args  is useless 
`bool` is not acceptable here

bancer replied:
Well, that's how it is used: `$this->render(false, $layout);`
https://github.com/cakephp/cakephp/blob/2.x/lib/Cake/Controller/Controller.php#L1017

And passed unaltered from Controller to View here - https://github.com/cakephp/cakephp/blob/2.x/lib/Cake/Controller/Controller.php#L963.

saeideng replied:
OK, so just `string|false` is good
and no needed to `!is_bool($view) &&`  (line 470)

garas commented on lib/Cake/Controller/Controller.php:
```diff
@@ -45,14 +45,21 @@
  * @property      EmailComponent $Email
  * @property      PaginatorComponent $Paginator
+ * @property      array $params Array of parameters parsed from the URL. Deprecated, use CakeRequest::$params instead.
+ * @property      string $here The full address to the current request. Deprecated, use CakeRequest::$here instead.
```
Sort properties alphabetically (`$here` and `$Flash` is out of order). And maybe list components first (also it will look better as they has no description here), then your additions.

bancer committed:
Improve documentation

bancer committed:
Improve documentation

bancer commented:
May this be merged already?
